cursor: let followMouse update focus only when cursor enters a surface

tokyo4j opened the pull request:
Fixes #2650.

This needs more testing and maybe this can be implemented in a cleaner way.

johanmalm commented:
LGTM.

I don't mind the old/new thing. Seems clean enough.

tokyo4j commented:
I noticed that we should update keyboard focus when entering a window including SSD rather than a surface... I'll make a follow-up PR later.

johanmalm commented:
Okay. Should be surface || view?
I did ponder that when reviewing. Silly me - should have said something.
Thanks for catching.

tokyo4j commented:
> Should be surface || view?
> I did ponder that when reviewing. Silly me - should have said something.

Yeah... I was kind of confused when writing this patch and I overlooked such a obvious difference. I admit, I'm not smart enough to understand how focus management works.

I'm now wondering how we should remember the cursor context of a previous motion handler call, which is necessary to detect whether the cursor is "entering" a window. It will take me some time to determine what's the best approach...

tokyo4j commented:
> Suggest trying the kwin approach if it simplifes things.

Just for clarification, labwc in the master branch currently have the same behavior as kwin that updates keyboard focus only when the cursor enters the content of a window. If we decide to follow kwin's behavior, all we have to do for the next release is just to describe the change in `NEWS.md` like:

```
CHANGED
- If <focus><followMouse> is set to yes, the focus is now updated only when the cursor enters
  a window content, not when the cursor moves within the window content, or when the cursor
  enters a window's titlebar (#2652).
```

ahesford commented:
I just ran into this behavior when I shaded a window and de-focused it, and found it impossible to refocus with the mouse. Because there is no surface to enter on a shaded window, you can only focus by clicking the titlebar.

In general, losing the ability to focus a window by mousing over the titlebar feels broken, but I don't know how long I'll put up with it before I decide to figure out a proper fix. This is one of those "we should probably fix this some day" things.
